auth/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+pub struct CredentialConfig {
+    /// The scopes that the minted [AccessToken] should have.
+    // TODO(codyoss): This should be optional.
+    pub scopes: Vec<String>,
+}
+
+/// A [AccessToken] producer that is automatically refreshed and can be shared across
+/// threads.
+pub struct Credential {
+    source: Box<dyn Source + Send + Sync + 'static>,
+}
+
+// TODO(codyoss): This is currently needed to make generated code easier to generate. Not
+//       great that it is an empty cred. Either this should do a partial ADC lookup
+//       or we should find a different way and remove this. If we did not need
+//       an api call for compute credentials this could just work. But today we
+//       need to make async network requests.
+impl Default for Credential {
+    fn default() -> Self {
+        Self {
+            source: Box::new(NoOpSource {}),
+        }
+    }
+}
+
+impl Credential {
+    /// Fetches a [AccessToken] based on environment and/or configuration settings.
+    pub async fn access_token(&self) -> Result<AccessToken> {
+        self.source.token().await
+    }
+
+    /// Creates a Credential that uses [Application Default Credentials](https://google.aip.dev/auth/4110)
+    /// to figure out how a to produce a [AccessToken].
+    // TODO(codyoss): maybe make this take a builder?
+    pub async fn find_default(config: CredentialConfig) -> Result<Credential> {
+        let base_source = Credential::base_source(config).await?;
+        let refreshed_source = RefresherSource {
+            source: base_source,
+            ..Default::default()
+        };
+        Ok(Credential {
+            source: Box::new(refreshed_source),
+        })
+    }
+
+    /// Finds a Source from which to create tokens.
+    async fn base_source(
+        config: CredentialConfig,
+    ) -> Result<Box<dyn Source + Send + Sync + 'static>> {
+        // 1: Known environment variable.
+        if let Ok(file_name) = std::env::var(GOOGLE_APPLICATION_CREDENTIALS) {
+            let source = Credential::file_source(file_name, config)?;
+            return Ok(source);
+        }
+        // 2: Well-known file.
+        if let Ok(path) = Credential::well_known_file() {
+            if path.exists() {
+                if let Some(file_name) = path.to_str() {
+                    let source = Credential::file_source(file_name.into(), config)?;
+                    return Ok(source);
+                }
+            }
+        }
+        // 3: Check if in an environment with on Google Cloud
+        if metadata::on_gce().await {
+            let source = ComputeSource::builder().scopes(config.scopes).build()?;
+            let source = Box::new(source);
+            return Ok(source);
+        }
+
+        Err(Error::Other("unable to detect default credentials".into()))
+    }
+
+    /// Creates a source from a file type credential such as a Service Account
+    /// Key file or a gcloud user credential.
+    fn file_source(
+        file_path: String,
```

Comment:
File systems can have file names that are not valid UTF-8, and therefore cannot be represented as a Rust String.
This code could never open such a file.  To fix, the type of `file_path` should match the type of the argument to `std::fs::read()`.  See https://doc.rust-lang.org/std/fs/fn.read.html.

Reply:
Switched to `PathBuf` for now, or would you prefer a `AsRef<Path>`?